added fix for compatibility with Debian Trixie

damisul opened the pull request:
I've recently updated my machine from Debian Bookworm to Debian Trixie and it turned out gem cannot be compiled due to changes in libc:
```
mmd/file.c: In function ‘absolute_path_for_argument’:
mmd/file.c:431:18: error: implicit declaration of function ‘realpath’ [-Wimplicit-function-declaration]
  431 |         result = realpath(arg, NULL);
      |                  ^~~~~~~~
mmd/file.c:431:16: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  431 |         result = realpath(arg, NULL);
      |                ^
```
It's been ~20 years since I've coded in C for the last time, but I've found an advice to add macro to enable compatibility with older standard:
```
#define _XOPEN_SOURCE 500
```
It worked for me.
Not sure if it will work for other platforms though, so I hope more expirienced C developers will review it.
@tillsc , FYI

tillsc commented:
Thanks for the PR!

`file.c` (and several other C files) are generated upstream by the MultiMarkdown-6 project, so patching them inside this gem would make upgrades hard to maintain.

I upgraded the bundled MultiMarkdown engine to version 6.7. While this release does not explicitly contain a realpath/glibc compatibility change, some generated files differ and may behave differently depending on the system headers.

Could you test whether the issue still occurs with rmultimarkdown 6.7 on your Trixie environment?

If it persists, I’m happy to explore a build-level fix (e.g. adding `_XOPEN_SOURCE` via compiler flags rather than patching generated code).

tillsc commented:
If 6.7 does not resolve the issue on your setup, you can help me verify the root cause by testing whether the extension builds successfully when `_XOPEN_SOURCE` is enabled manually:

    CFLAGS="-D_XOPEN_SOURCE=500" bundle install

or, when installing the gem directly:

    CFLAGS="-D_XOPEN_SOURCE=500" gem install rmultimarkdown

If the build works with this flag but fails without it, please let me know. In that case I will add the flag directly in `extconf.rb` so users on glibc systems won’t have to apply workarounds.

damisul commented:
6.7 does not works out of the box

> If 6.7 does not resolve the issue on your setup, you can help me verify the root cause by testing whether the extension builds successfully when `_XOPEN_SOURCE` is enabled manually:
> 
> ```
> CFLAGS="-D_XOPEN_SOURCE=500" bundle install
> ```
> or, when installing the gem directly:
> 
> ```
> CFLAGS="-D_XOPEN_SOURCE=500" gem install rmultimarkdown
> ```

Those two options does not works too, still same error:
```
compiling mmd/file.c
mmd/file.c: In function ‘absolute_path_for_argument’:
mmd/file.c:441:18: error: implicit declaration of function ‘realpath’ [-Wimplicit-function-declaration]
  441 |         result = realpath(arg, NULL);
      |                  ^~~~~~~~
mmd/file.c:441:16: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  441 |         result = realpath(arg, NULL);
      |                ^
At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make: *** [Makefile:248: mmd/file.o] Error 1

make failed, exit code 2

```

tillsc commented:
Could you modify [this line in extconf.rb](https://github.com/tillsc/multi_markdown/blob/d76265d5462dc2b110f71e70d7942a25d1b1542b/ext/extconf.rb#L14)

    $CFLAGS << '  -DNDEBUG=1 -DDISABLE_OBJECT_POOL'

so it becomes:

    $CFLAGS << '  -DNDEBUG=1 -DDISABLE_OBJECT_POOL -D_XOPEN_SOURCE=700'

and make the same change to the $CXXFLAGS line as well?

Then rebuild locally (e.g. `rake clean && rake`) and let me know if the build succeeds. If this fixes it on your system, I’ll add the flag for glibc-based builds.

damisul committed:
Added -D_XOPEN_SOURCE=700 to compiler flags for compatibility with newer linux distos (like Debian Trixie)
Bumped version to 6.7.0.1

damisul commented:
> so it becomes:

> $CFLAGS << '  -DNDEBUG=1 -DDISABLE_OBJECT_POOL -D_XOPEN_SOURCE=700'

> and make the same change to the $CXXFLAGS line as well?

Hi @tillsc . Yes, this works!

I've updated PR to include those changes (also rebased branch to current master to incorporate upgrade to 6.7)

damisul commented:
I can confirm it compiles well both in Debian Bookworm (libc 2.36 / gcc 12.2)  and in Debian Trixie (libc 2.41 / gcc 14.2)

tillsc commented:
Thanks for the follow-up!

I’ve added a Linux-only guard and switched the build flag to `-D_XOPEN_SOURCE=500` (the minimal value that fixes the issue) in the extension build configuration.

Because of that, I won’t merge the PR touching generated files. Instead, I’ve released rmultimarkdown 6.7.0.1, which should build cleanly on your system now.

Could you please confirm that this works for you? If so, you can close the issue / PR.

damisul commented:
@tillsc , I confirm it works for us now. Thank you!
